Sort hierarchy items deterministically

NobodysNightmare opened the pull request:
The lists were a kind of make-shift solution to allow using hierarchy custom fields in all of our UIs. However, the sort order coming out of the service was rather useless to a user. By sorting, we make dropdowns resulting from the returned items easier to navigate.

# Ticket
* Implementation: https://community.openproject.org/wp/67454
* Feature request: https://community.openproject.org/wp/59173

# Sorting difference

This is the sorting difference for custom fields on my local machine:

```
# no sorting
Lizu Sapa
Milky Way
Milky Way / Sol
Lizu Sapa / Toril
Milky Way / Alpha Centauri
Milky Way / Alpha Centauri / Proxima Centauri
Lizu Sapa / Toril / Selûne
Milky Way / Sol / Mercury (A)
Milky Way / Sol / Venus (B)
Lizu Sapa / Toril / Tears of Selûne
Milky Way / Sol / Earth (C)
Milky Way / Sol / Mars (D)
Milky Way / Sol / Jupiter (E)
Milky Way / Sol / Saturn (F)
Milky Way / Sol / Uranus (G)
Milky Way / Sol / Neptune (H)
Milky Way / Sol / Pluto
Milky Way / Sol / Earth (C) / Moon (I)
Milky Way / Sol / Jupiter (E) / Io (I)
Milky Way / Sol / Saturn (F) / Titan
Milky Way / Sol / Pluto / Charon
Milky Way / Sol / Mars (D) / Phobos
Milky Way / Sol / Jupiter (E) / Europa (II)
Milky Way / Sol / Pluto / Styx
Milky Way / Sol / Mars (D) / Deimos
Milky Way / Sol / Jupiter (E) / Ganymede (III)
Milky Way / Sol / Pluto / Nix
Milky Way / Sol / Pluto / Kerberos
Milky Way / Sol / Jupiter (E) / Callisto (IV)
Milky Way / Sol / Pluto / Hydra

# preordered
Lizu Sapa
Lizu Sapa / Toril
Lizu Sapa / Toril / Selûne
Lizu Sapa / Toril / Tears of Selûne
Milky Way
Milky Way / Sol
Milky Way / Sol / Mercury (A)
Milky Way / Sol / Venus (B)
Milky Way / Sol / Earth (C)
Milky Way / Sol / Earth (C) / Moon (I)
Milky Way / Sol / Mars (D)
Milky Way / Sol / Mars (D) / Phobos
Milky Way / Sol / Mars (D) / Deimos
Milky Way / Sol / Jupiter (E)
Milky Way / Sol / Jupiter (E) / Io (I)
Milky Way / Sol / Jupiter (E) / Europa (II)
Milky Way / Sol / Jupiter (E) / Ganymede (III)
Milky Way / Sol / Jupiter (E) / Callisto (IV)
Milky Way / Sol / Saturn (F)
Milky Way / Sol / Saturn (F) / Titan
Milky Way / Sol / Uranus (G)
Milky Way / Sol / Neptune (H)
Milky Way / Sol / Pluto
Milky Way / Sol / Pluto / Charon
Milky Way / Sol / Pluto / Styx
Milky Way / Sol / Pluto / Nix
Milky Way / Sol / Pluto / Kerberos
Milky Way / Sol / Pluto / Hydra
Milky Way / Alpha Centauri
Milky Way / Alpha Centauri / Proxima Centauri
```

NobodysNightmare committed:
Sort hierarchy items by the name they are displayed with

The lists were a kind of make-shift solution to allow
using hierarchy custom fields in all of our UIs. However,
the sort order coming out of the service was rather useless
to a user. By sorting by the ancestry path, we sort the list by
the string that's later visible to the user, thus making
it easier to navigate the corresponding dropdown.

NobodysNightmare commented on app/forms/custom_fields/inputs/multi_select_list.rb:
```diff
@@ -82,6 +82,7 @@ def hierarchical_list_items
     CustomFields::Hierarchy::HierarchicalItemService.new
       .get_descendants(item: @custom_field.hierarchy_root, include_self: false)
       .value_or([])
+      .sort_by { |item| item.ancestry_path }
```
This solution was chosen because it's pretty local to the forms where we see the problem and it works exactly as it is useful for here.

Alternatively we could try to change `HierarchicalItemService#get_descendants` to return a defined order. Something that I noticed is that it is possible to sort the children of an item manually. So it would be arguable that `get_descendants` should return items in a defined order. What would fit our purpose here:

* return items depth first
* order them according to their position relative to their siblings

This might yield better results for hierarchies that are consciously sorted, e.g. `low, medium, high`, which would be sorted `high, low, medium` alphabetically.

On the other hand, the sorting would be applied in more places, maybe even when no sorting is needed.

NobodysNightmare replied:
I now implemented this alternative suggested here in commit 2.

NobodysNightmare commented:
@Kharonus Maybe I'll have to look at this tomorrow once more. The solution here definitely works and is an extension of previous code that was specifically prepared for hierarchy custom fields. So it can't be too wrong.

However, I just noticed that we have a much nicer UI for hierarchy custom fields in the work package edit UI. And that is also Angular-based, so I don't quite know why we are not using it here...

<img width="829" height="331" alt="image" src="https://github.com/user-attachments/assets/9d7ff0ed-4345-48b5-98dd-1a341994a3a6" />

Kharonus commented:
I know. The reason is, that the input fields in the work package view are quite heavily tied to the work package resource and API requests. We can try, but there is a reason I wasn't able to use it easily in e.g. the bulk edit.

We can have a look together, but my bet is, that it isn't trivial.

NobodysNightmare committed:
Make result order returned by get_descendants deterministic

Previously the results were not returned in any specific order,
making the result not immediately usable in flat lists of the hierarchy.

By preordering the result, it's not necessary anymore to apply later
alphabetical ordering or similar. This is preferable, because alphabetical
ordering was only a workaround that didn't take specific ordering established
by the user into account.
